os/board/rtl8721csm/src/component/common/bluetooth/realtek/sdk/example/ble_peripheral/ble_tizenrt_server.c

```diff
@@ -536,13 +536,16 @@ trble_result_e rtw_ble_server_disconnect(trble_conn_handle con_handle)
         return TRBLE_INVALID_STATE;
     }
 
-    if (con_handle != 0)
+    trble_conn_handle *conn_id = os_mem_alloc(0, sizeof(trble_conn_handle));
+    if(conn_id == NULL)
     {
-        return TRBLE_NOT_FOUND;
+        debug_print("\n[%s] Memory allocation failed \n");
+        return TRBLE_FAIL;
     }
-
-    if(ble_tizenrt_server_send_msg(BLE_TIZENRT_MSG_DISCONNECT, NULL) == false)
+    *conn_id = con_handle;
+    if(ble_tizenrt_server_send_msg(BLE_TIZENRT_MSG_DISCONNECT, conn_id) == false)
     {
+        os_mem_free(conn_id);
```

Comment:
If the `if` is false, then `conn_id` will be free. How about a true case? Who free the allocated memory?

Reply:
@sunghan-chang if it is true, then the pointer conn_id is passed to ble_tizenrt_scatternet_handle_upstream_msg, and it is released at the end of the ble_tizenrt_scatternet_handle_upstream_msg function.

Reply:
Is this passed to `ble_tizenrt_scatternet_handle_upstream_msg`? not `ble_tizenrt_app_handle_upstream_msg`?
I searched like below
1. find `ble_tizenrt_server_send_msg` -> This is assigned to `ble_tizenrt_send_msg` at ble_tizenrt_server.c line 38.
2. `ble_tizenrt_send_msg` sends two messages using `ble_tizenrt_io_queue_handle` and `ble_tizenrt_evt_queue_handle`.
3. Receiver of them is `ble_tizenrt_app_main_task` and it calls `ble_tizenrt_app_handle_io_msg`
4. It calls `ble_tizenrt_app_handle_upstream_msg` for `IO_MSG_TYPE_QDECODE`
5. `ble_tizenrt_app_handle_upstream_msg` does not free the buffer.

Could you check what is missing?

Reply:
Yes it is passed to `ble_tizenrt_scatternet_handle_upstream_msg`.

1. Tizenrt uses BT combo mode, which `ble_tizenrt_server_send_msg `-> is assigned to `ble_tizenrt_scatternet_send_msg `in `rtw_ble_combo_init`.
2. `ble_tizenrt_scatternet_send_msg  `sends the message with `ble_tizenrt_scatternet_io_queue_handle `and `ble_tizenrt_scatternet_evt_queue_handle`
3. And then it will be handled by `ble_tizenrt_scatternet_app_main_task`, which will lead to `ble_tizenrt_scatternet_app_handle_io_msg`
4. Finally `ble_tizenrt_scatternet_handle_upstream_msg` will handle it and release the memory.

Reply:
Oh I missed the initialization function. Because of the config (supporting scatternet), it calls rtw_ble_combo_init, not rtw_ble_server_init. Thank you for explanation.